core/utils/utils.go

```diff
@@ -437,17 +437,37 @@ type StopChan chan struct{}
 
 // NewCtx returns a background [context.Context] that is cancelled when StopChan is closed.
 func (s StopChan) NewCtx() (context.Context, context.CancelFunc) {
-	return s.Ctx(context.Background())
+	return StopRChan((<-chan struct{})(s)).NewCtx()
 }
 
 // Ctx cancels a [context.Context] when StopChan is closed.
 func (s StopChan) Ctx(ctx context.Context) (context.Context, context.CancelFunc) {
-	return s.CtxCancel(context.WithCancel(ctx))
+	return StopRChan((<-chan struct{})(s)).Ctx(ctx)
 }
 
 // CtxCancel cancels a [context.Context] when StopChan is closed.
 // Returns ctx and cancel unmodified, for convenience.
 func (s StopChan) CtxCancel(ctx context.Context, cancel context.CancelFunc) (context.Context, context.CancelFunc) {
+	return StopRChan((<-chan struct{})(s)).CtxCancel(ctx, cancel)
+}
+
+// A StopRChan signals when some work should stop.
+// This version is receive-only.
+type StopRChan <-chan struct{}
```

Comment:
This parallel type seems odd, but I didn' t see a simpler way to support cases where you don't own the original chan, e.g. `context.Done()` returns a receive-only `<-chan struct { }`.